airbyte-cdk/bulk/toolkits/extract-cdc/src/main/kotlin/io/airbyte/cdk/read/cdc/CdcPartitionReader.kt

```diff
@@ -0,0 +1,224 @@
+                // TODO : send close analytics message
+                Thread({ engine.close() }, "debezium-close").start()
+            }
+        }
+    }
+
+    inner class CompletionCallback : DebeziumEngine.CompletionCallback {
+        override fun handle(success: Boolean, message: String?, error: Throwable?) {
+            if (success) {
+                log.info { "Debezium engine has shut down successfully: $message" }
+            } else {
+                // There are cases where Debezium doesn't succeed but only fills the message field.
+                val e: Throwable = error ?: RuntimeException(message)
+                log.warn(e) { "Debezium engine has NOT shut down successfully: $message" }
+                throw e
+            }
+        }
+    }
+
+    inner class ConnectorCallback : DebeziumEngine.ConnectorCallback {
+        override fun connectorStarted() {
+            log.info { "Debezium connector started" }
+        }
+
+        override fun connectorStopped() {
+            log.info { "Debezium connector stopped" }
+        }
+
+        override fun taskStarted() {
+            log.info { "Debezium task started" }
+        }
+
+        override fun taskStopped() {
+            log.info { "Debezium Task stopped" }
+        }
+    }
+
+    enum class CloseReason(val message: String) {
+        TIMEOUT("timed out"),
+        HEARTBEAT_REACHED_TARGET_POSITION(
+            "heartbeat indicates that WAL consumption has reached the target position"
+        ),
+        RECORD_REACHED_TARGET_POSITION(
+            "record indicates that WAL consumption has reached the target position"
+        ),
+    }
+
+    companion object {
+
+        val debeziumVersion: String = DebeziumEngine::class.java.getPackage().implementationVersion
+
+        /**
+         * Extracts the [SourceRecord] wrapped inside a Debezium [ChangeEvent].
+         *
+         * [sourceRecordField] acts as a cache so that we avoid using reflection at each lookup.
+         */
+        fun getSourceRecord(event: ChangeEvent<String?, String?>): SourceRecord? {
+            if (!embeddedEngineChangeEventClass.isInstance(event)) {
+                // This is very unlikely, but we should guard against it.
+                // We don't control Debezium internals
+                return null
+            }
+            val eventClass: Class<out ChangeEvent<*, *>?> = event::class.java
+            val f: Field =
+                sourceRecordField.getOrPut(eventClass) {
+                    eventClass.getDeclaredField("sourceRecord").apply { isAccessible = true }
+                }
+            return f[event] as? SourceRecord // Again, this is very unlikely to be null.
+        }
+
+        private val sourceRecordField = ConcurrentHashMap<Class<out ChangeEvent<*, *>>, Field>()
+        private val embeddedEngineChangeEventClass: Class<*> =
+            Class.forName("io.debezium.embedded.EmbeddedEngineChangeEvent")
```

Comment:
Your concern is valid. I'm not sure why we're not using `EmbeddedEngineChangeEvent::class.java` directly. This would go a long way to making `getSourceRecord` less brittle. Perhaps this wasn't possible in older versions of Debezium? In any case it's public now. Also, the logic in `getSourceRecord` is needlessly convoluted. I'll adjust.

I'll also add an assertion in the test that checks that the `SourceRecord` is always available.

The debezium dependency is explicitly only on this CDK toolkit, not the whole CDK. I think that's OK, that's the kind of thing toolkits are for.

Reply:
Added some statistics in `CdcRecordReader::run` along with some assertions.